File: src/main/java/com/innatour/toursmanager/controller/AuthController.java
package com.innatour.toursmanager.controller;

import com.innatour.toursmanager.dto.JwtResponse;
import com.innatour.toursmanager.dto.LoginRequest;
import com.innatour.toursmanager.dto.RegisterRequest;
import com.innatour.toursmanager.model.Guide;
import com.innatour.toursmanager.repository.GuideRepository;
import com.innatour.toursmanager.security.JwtTokenProvider;
import com.innatour.toursmanager.service.GuideService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

/**
 * Authentication Controller
 * 
 * Handles:
 * - POST /api/auth/register - Register new guide
 * - POST /api/auth/login - Login and get JWT token
 */
@RestController
@RequestMapping("/api/auth")
@CrossOrigin(origins = "*")
@Tag(name = "Authentication", description = "Guide authentication APIs - Register and login")
public class AuthController {
    
    private final GuideService guideService;
    private final GuideRepository guideRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;
    
    public AuthController(GuideService guideService, 
                         GuideRepository guideRepository,
                         PasswordEncoder passwordEncoder, 
                         JwtTokenProvider jwtTokenProvider) {
        this.guideService = guideService;
        this.guideRepository = guideRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtTokenProvider = jwtTokenProvider;
    }
    
    /**
     * Register a new guide
     * POST /api/auth/register
     */
    @Operation(summary = "Register new guide", description = "Create a new guide account with email and password")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "201", description = "Guide successfully registered",
                content = @Content(mediaType = "application/json", schema = @Schema(implementation = JwtResponse.class))),
        @ApiResponse(responseCode = "400", description = "Email already exists or invalid input", content = @Content)
    })
    @PostMapping("/register")
    public ResponseEntity<?> register(@Valid @RequestBody RegisterRequest registerRequest) {
        // Check if email already exists
        if (guideRepository.findByEmail(registerRequest.getEmail()).isPresent()) {
            return ResponseEntity.badRequest().body("Email already in use");
        }
        
        // Create guide entity
        Guide guide = new Guide();
        guide.setFirstName(registerRequest.getFirstName());
        guide.setLastName(registerRequest.getLastName());
        guide.setEmail(registerRequest.getEmail());
        guide.setPhoneNumber(registerRequest.getPhoneNumber());
        guide.setProfile(registerRequest.getProfile());
        guide.setActive(registerRequest.getActive() != null ? registerRequest.getActive() : true);
        
        // Hash password with BCrypt
        guide.setPassword(passwordEncoder.encode(registerRequest.getPassword()));
        
        // Create guide with languages
        Guide createdGuide = guideService.createGuide(guide, registerRequest.getLanguages());
        
        // Generate JWT token
        String token = jwtTokenProvider.generateToken(createdGuide.getEmail());
        
        // Return token and user info
        JwtResponse jwtResponse = new JwtResponse(
                token,
                createdGuide.getEmail(),
                createdGuide.getFirstName(),
                createdGuide.getLastName()
        );
        
        return ResponseEntity.status(HttpStatus.CREATED).body(jwtResponse);
    }
    
    /**
     * Login
     * POST /api/auth/login
     */
    @Operation(summary = "Login", description = "Authenticate guide and receive JWT token")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "Successfully authenticated",
                content = @Content(mediaType = "application/json", schema = @Schema(implementation = JwtResponse.class))),
        @ApiResponse(responseCode = "401", description = "Invalid credentials", content = @Content)
    })
    @PostMapping("/login")
    public ResponseEntity<?> login(@Valid @RequestBody LoginRequest loginRequest) {
        // Find guide by email
        Guide guide = guideRepository.findByEmail(loginRequest.getEmail())
                .orElseThrow(() -> new IllegalArgumentException("Invalid email or password"));
        
        // Verify password
        if (!passwordEncoder.matches(loginRequest.getPassword(), guide.getPassword())) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid email or password");
        }
        
        // Generate JWT token
        String token = jwtTokenProvider.generateToken(guide.getEmail());
        
        // Return token and user info
        JwtResponse jwtResponse = new JwtResponse(
                token,
                guide.getEmail(),
                guide.getFirstName(),
                guide.getLastName()
        );
        
        return ResponseEntity.ok(jwtResponse);
    }
}

File: dotnet-api/ToursManager.Api/Controllers/AuthController.cs
using Microsoft.AspNetCore.Mvc;
using ToursManager.Api.DTOs;
using ToursManager.Api.Models;
using ToursManager.Api.Services;

namespace ToursManager.Api.Controllers;

/// <summary>
/// Authentication Controller - handles register and login.
/// Equivalent to Spring Boot's @RestController.
/// </summary>
[ApiController]  // Enables automatic model validation and binding
[Route("api/[controller]")]  // Maps to /api/auth
public class AuthController : ControllerBase
{
    private readonly GuideService _guideService;
    private readonly JwtService _jwtService;
    private readonly ILogger<AuthController> _logger;
    
    public AuthController(
        GuideService guideService,
        JwtService jwtService,
        ILogger<AuthController> logger)
    {
        _guideService = guideService;
        _jwtService = jwtService;
        _logger = logger;
    }
    
    /// <summary>
    /// Register a new guide
    /// POST /api/auth/register
    /// </summary>
    /// <response code="201">Guide successfully registered</response>
    /// <response code="400">Email already exists or invalid input</response>
    [HttpPost("register")]
    [ProducesResponseType(typeof(JwtResponse), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<ActionResult<JwtResponse>> Register([FromBody] RegisterRequest request)
    {
        // Check if email already exists
        if (await _guideService.EmailExistsAsync(request.Email))
        {
            return BadRequest(new { message = "Email already in use" });
        }
        
        // Create guide entity
        var guide = new Guide
        {
            FirstName = request.FirstName,
            LastName = request.LastName,
            Email = request.Email,
            PhoneNumber = request.PhoneNumber,
            Profile = request.Profile,
            Active = request.Active,
            // Hash password with BCrypt
            Password = BCrypt.Net.BCrypt.HashPassword(request.Password)
        };
        
        // Handle languages if provided
        if (!string.IsNullOrWhiteSpace(request.Languages))
        {
            var codes = request.Languages.Split(',', StringSplitOptions.RemoveEmptyEntries)
                .Select(c => c.Trim().ToLower())
                .ToList();
            
            var languages = await _guideService.GetLanguagesByCodesAsync(codes);
            guide.Languages = languages;
        }
        
        // Save guide
        var createdGuide = await _guideService.CreateAsync(guide);
        
        // Generate JWT token
        var token = _jwtService.GenerateToken(createdGuide.Email);
        
        // Return token and user info
        var response = new JwtResponse(
            token,
            createdGuide.Email,
            createdGuide.FirstName,
            createdGuide.LastName
        );
        
        return CreatedAtAction(nameof(Register), response);
    }
    
    /// <summary>
    /// Login - authenticate and receive JWT token
    /// POST /api/auth/login
    /// </summary>
    /// <response code="200">Login successful</response>
    /// <response code="401">Invalid credentials</response>
    [HttpPost("login")]
    [ProducesResponseType(typeof(JwtResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    public async Task<ActionResult<JwtResponse>> Login([FromBody] LoginRequest request)
    {
        // Find guide by email
        var guide = await _guideService.GetByEmailAsync(request.Email);
        if (guide == null)
        {
            return Unauthorized(new { message = "Invalid email or password" });
        }
        
        // Verify password
        if (!BCrypt.Net.BCrypt.Verify(request.Password, guide.Password))
        {
            return Unauthorized(new { message = "Invalid email or password" });
        }
        
        // Generate JWT token
        var token = _jwtService.GenerateToken(guide.Email);
        
        // Return token and user info
        var response = new JwtResponse(
            token,
            guide.Email,
            guide.FirstName,
            guide.LastName
        );
        
        return Ok(response);
    }
}
